### PR TITLE
coprocessor: implemented FIELD() function

### PR DESCRIPTION
**Pull request by manifoldQAQ**
## What have you changed? (mandatory)

Implemented `FIELD()` function in MySQL, which correspond to the three functions `FieldInt`, `FieldReal`, `FieldString`.

## What are the type of the changes? (mandatory)

New feature.

## How has this PR been tested? (mandatory)

Unit tests as in the source code.

## Does this PR affect documentation (docs) update? (mandatory)

Yes. Please mark this function as resolved in #3275 

## Does this PR affect tidb-ansible update? (mandatory)

No.

### CODE REVIEW

**Comment by manifoldQAQ** [comment]
PTAL. It seems that there's something wrong with CI.

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 1]
```diff
@@ -1572,7 +1613,7 @@ mod tests {
         let cases = vec![
             (
                 vec![
-		    Datum::Bytes(b",".to_vec()),
+            Datum::Bytes(b",".to_vec()),
```
Comment (rleungx) [inline src/coprocessor/dag/expr/builtin_string.rs]: This indent is wired.

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 2]
```diff
@@ -312,6 +312,47 @@ impl ScalarFunc {
         self.children[i as usize].eval_string(ctx, row)
     }
 
+    #[inline]
+    pub fn field_int(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        // Pattern matching made explicit here.
+        // As per the MySQL doc, if the first argument is NULL, this function always returns 0.
+        let s = match self.children[0].eval_string(ctx, row) {
```
Comment (huachaohuang) [inline src/coprocessor/dag/expr/builtin_string.rs]: I think we should `eval_int` here?
BTW, we can do it like `match self.children[0].eval_int(ctx, row)?`.

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 3]
```diff
@@ -312,6 +312,47 @@ impl ScalarFunc {
...
+            Err(e) => return Err(e),
+            Ok(None) => return Ok(Some(0)),
+            Ok(Some(res)) => res,
+        };
+
+        let mut i = 1;
+        while i < self.children.len() as i64 {
```
Comment (huachaohuang) [inline src/coprocessor/dag/expr/builtin_string.rs]: How about `self.children.iter().skip(1).enumerate()`?

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 4]
```diff
@@ -312,6 +312,47 @@ impl ScalarFunc {
...
+        if i == self.children.len() as i64 {
+            Ok(Some(0))
+        } else {
+            Ok(Some(i))
+        }
+    }
+
+    #[inline]
+    pub fn field_real(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<f64>> {
+        Ok(Some(try_opt!(self.field_int(ctx, row)) as f64))
```
Comment (huachaohuang) [inline src/coprocessor/dag/expr/builtin_string.rs]: We should `eval_real` here.

File: src/coprocessor/dag/expr/builtin_string.rs (hunk 2 above)
Comment (manifoldQAQ) [reply src/coprocessor/dag/expr/builtin_string.rs]: Nope. All the arguments of the `FIELD()` function are strings as per [MySQL doc](https://dev.mysql.com/doc/refman/8.0/en/string-functions.html#function_field).

Comment (manifoldQAQ) [reply src/coprocessor/dag/expr/builtin_string.rs]: It seems that I misunderstood the meaning. I'll fix that.

File: src/coprocessor/dag/expr/builtin_string.rs (hunk 3 above)
Comment (manifoldQAQ) [reply src/coprocessor/dag/expr/builtin_string.rs]: It seems hard to model this function using only combinators due to subtle handling of NULL. I used a for loop instead.

File: src/coprocessor/dag/expr/builtin_string.rs (hunk 4 above)
Comment (manifoldQAQ) [reply src/coprocessor/dag/expr/builtin_string.rs]: Fixed.

**Comment by manifoldQAQ** [comment]
@huachaohuang PTAL

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 5]
```diff
@@ -1854,6 +1902,77 @@ mod tests {
         }
     }
 
+    #[test]
+    fn test_field_string() {
```
Comment (huachaohuang) [inline src/coprocessor/dag/expr/builtin_string.rs]: Please add more tests for `field_int` and `field_real`.

Comment (manifoldQAQ) [reply src/coprocessor/dag/expr/builtin_string.rs]: @huachaohuang Done.

**Comment by manifoldQAQ** [comment]
There's something wrong w/ the CI. Please force a rebuild.

**Pull request merged**